### DISCUSSION TITLE
Problems with Bioawk commands in jupyter

### DISCUSSION THREAD

**Discussion by jdduprey** [sr320/course-fish546-2021#49] [Q&A]
`bioawk -cfastx 'END{print NR}' Good-lib1_S1_L001_R1_001.fastq`
This line of code returns a number of sequences when I run it in the terminal. But when I run it in jupyter:

`%%bash
bioawk -cfastx 'END{print NR}' ../data/Good-lib1_S1_L001_R1_001.fastq`

`bash: line 1: bioawk: command not found`
`CalledProcessError: Command 'b"bioawk -cfastx 'END{print NR}' ../data/Good-lib1_S1_L001_R1_001.fastq\n"' returned non-zero exit status 127.`

**Comment by kubu4** [comment]
Two things to resolve this:

1. `%%bash bioawk -cfastx 'END{print NR}' ../data/Good-lib1_S1_L001_R1_001.fastq`: This should be on two lines, like so:

```
%%bash
bioawk -cfastx 'END{print NR}' ../data/Good-lib1_S1_L001_R1_001.fastq
```

I'm a bit surprised you didn't get this error: `UsageError: %%bash is a cell magic, but the cell body is empty.`

2. Use the absolute path to call `bioawk`. If you're going to repeatedly run `bioawk` with the full path, you probably should set the path to a variable. Have a cell where you do this (and don't forget to run the cell; you'll need to re-run this cell to reload the variables any time you've closed the notebook and reopen it). Here's an example of how:

```
# Programs
%env seqtk=/home/samb/programs/seqtk_1.3-r115/seqtk
```

Then, in any subsequent cell were you need to run your program (in the example, `seqtk`), you refer to the variable in this fashion:

```
# Run seqtk
${seqtk} sam.fasta
```

Using this approach has potential to save you a lot of typing. Also, it makes the notebook more "portable", in that if you wanted to use your notebook on a different computer, you would just need to update a single cell with a different `bioawk` location.

---

This should get you up and running, but I haven't gone into _why_ you can't just call `bioawk`. The answer to that could be annoyingly long and complicated, as there are numerous things involved here (how was `bioawk` installed, how are you launching Jupyter Notebook, examining your system $PATH, etc).

**Reply by jdduprey** [reply]
Hey Sam, sorry my code block was formatted badly, next time I'll link the actual notebook like Steven suggests. I'll try the path solution you suggest. 

I installed bioawk with Homebrew.

**Comment by sr320** [comment]
Also it is always a good idea to provide a URL to Jupyter notebook on GitHub so folks can see exactly how the problem manifests.
